integration-cli/docker_cli_run_test.go

```diff
@@ -4307,3 +4307,63 @@ func (s *DockerSuite) TestRunTooLongHostname(c *check.C) {
 	hostname3 := "this-is-a-hostname-with-64-bytes-so-will-not-give-an-error.local"
 	dockerCmd(c, "run", "--hostname", hostname3, "busybox", "echo", "test")
 }
+
+func (s *DockerSuite) TestRunRmWithRestartPolicy(c *check.C) {
+	// run container test1 with restart policy "always", it won't be removed until stoped
+	dockerCmd(c, "run", "-d", "--name=test1", "--rm", "--restart=always", "busybox", "false")
+	// run container test2 with restart policy "on-failure", it will be autoremoved
+	dockerCmd(c, "run", "-d", "--name=test2", "--rm", "--restart=on-failure:3", "busybox", "false")
+
+	// wait until test2 is auto removed.
+	waitTime := 10 * time.Second
+	if daemonPlatform == "windows" {
+		// nslookup isn't present in Windows busybox. Is built-in.
+		waitTime = 120 * time.Second
+	}
+
+	err := waitRemoved("test2", waitTime)
+	c.Assert(err, checker.IsNil, check.Commentf("test2 should be removed in timeout"))
+
+	time.Sleep(2 * time.Second)
+	// listening events, check restart count
+	var (
+		startCount int
+		destroyed  bool
+	)
+	out, _ := dockerCmd(c, "events", "--since=0", fmt.Sprintf("--until=%d", daemonTime(c).Unix()))
+	events := strings.Split(strings.TrimSpace(out), "\n")
+
+	nEvents := len(events)
+	c.Assert(nEvents, checker.GreaterOrEqualThan, 1) //Missing expected event
+	actions := eventActionsByIDAndType(c, events, "test1", "container")
+
+	for _, a := range actions {
+		switch a {
+		case "start":
+			startCount++
+		case "destroy":
+			destroyed = true
+		}
+	}
+	c.Assert(startCount, checker.GreaterThan, 1, check.Commentf("test1 should have restarted more than once without auto removed: %v", actions))
+	c.Assert(destroyed, checker.False, check.Commentf("test1 shouldn't be destroyed: %v", actions))
+
+	actions = eventActionsByIDAndType(c, events, "test2", "container")
+	startCount = 0
+	destroyed = false
+	for _, a := range actions {
+		switch a {
+		case "start":
+			startCount++
+		case "destroy":
+			destroyed = true
+		}
+	}
+	c.Assert(startCount, checker.Equals, 4, check.Commentf("test2 should have restarted 3 times: %v\n%s", actions, out))
+	c.Assert(destroyed, checker.True, check.Commentf("test2 should be destroyed: %v\n%s", actions, out))
+
+	// stop test1, check if it's autoremoved
+	dockerCmd(c, "stop", "test1")
+	err = waitRemoved("test1", waitTime)
+	c.Assert(err, checker.IsNil, check.Commentf("test1 should be auto removed after manually stopped"))
+}
```

Comment:
Can you add a test that does:
- `docker run --name=test --rm -d busybox sh -c "sleep 5 ; exit 2"`
- `docker wait test`
- verifies that wait prints "2" and has an exit code of zero.
I'd like to make sure that wait still gets the right container exit code even when -d and --rm are used.
Unless this is already tested and I missed it.